### ISSUE TITLE
[New rule] Avoid Ignoring Return Values

### ISSUE THREAD

**Issue by vHanda** [dart-code-checker/dart-code-metrics#364]
**Please describe what the rule should do:**
We should never silently ignore a return value of a function. 

**If your rule is inspired by other please provide link to it:**

- Swift has a [similar feature](https://www.avanderlee.com/swift/discardableresult/)
- Request for this in [Typescript](https://github.com/microsoft/TypeScript/issues/8584)
- Java Jetbrains issue for such [an annotation](https://youtrack.jetbrains.com/issue/KT-12719)
- Golang [proposal](https://github.com/golang/go/issues/20803)
- GoLang IDE (Goland by Jetbrains) implements [such a warning](https://travix.io/unhandled-errors-in-go-3f341f2704dd)
- Rust [#must_use](https://doc.rust-lang.org/std/result/#results-must-be-used)
- OCaml warns about unused return values by default. It provides an [ignore](https://ocaml.org/api/Pervasives.html) function in the standard library.

I found out about this from this wonderful [blog post by Joe Duffy](http://joeduffyblog.com/2016/02/07/the-error-model/) (Warning: Super long)

**What category of rule is this? (place an "X" next to just one item)**

[X] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about (it will be better if you can provide both good and bad examples):**

```dart
int foo() {
    return 5;
}

void goo() {
    print('whatever');
}

void main() {
    goo(); // fine, no return value
    var t = foo(); // fine, return value is assigned
    foo(); // BAD: return value is silently ignored

    var str = "Hello there";
    str.substr(5); // BAD: Strings are immutable and the return value should be handled

    var date = new DateTime(2018, 1, 13);
    date.add(Duration(days: 1, hours: 23))); // BAD: Return value ignored, DateTime is immutable
}
```

**The rationale behind the rule:**

* Immutable classes: DateTime/String are immutable and operations on them return a new object. One sometimes forgets about this and bugs are introduced.

* Some people are using the [Result](https://pub.dev/packages/result) class and avoiding throwing exceptions. Here each operation returns a union type with either the expected result or an error. Sometimes, certain operations only return an error. It would be nice to be reminded to handle that error.

  I've been using this pattern in [dart-git](https://github.com/GitJournal/dart-git) where I have found it is critical to handle all errors. Exceptions haven't worked for me, as we don't have checked-exceptions.

**Are you willing to submit a pull request to implement this rule?** No 

(maybe? I haven't ever implemented a rule before, and I'm not sure how long it will take me, though I'm willing to try - but I don't want to promise anything)

**Comment by incendial** [comment]
@vHanda thanks for suggesting this rule. It looks awesome and definitely nice to have, but I don't think that we can easily get the return type of an invocation while performing static analysis. I will take a closer look and return with the results. 

Dart team recently added a new annotation to meta package [called `useResult`](https://github.com/dart-lang/sdk/blob/master/pkg/meta/lib/meta.dart#L287). I think that they try to solve the same problem, although it will require all libraries to add it to their source code.

**Comment by incendial** [comment]
Nope, I was wrong, we have a return type info! Seems that there are blockers for implementing this rule.

**Comment by incendial** [comment]
@vHanda have you though about exceptions for this rule? I think that `void`, `Never`, `Null`, `Future<void>`, `Future<Null>` should be definitely ignored, but maybe there is more?

And how do you think we should handle methods that mutate the data and return something (and you might not want to handle the returned value)?

**Comment by incendial** [comment]
> If possible, could you please point me towards a similar rule I could look at to implement this?

I'm not sure that we have a similar one. But you need to implement `visitMethodInvocation` in a visitor and check if an invocation is **not** used in an assignment or cascade and **not** passed as an argument (maybe there are more edge cases). And have a not exceptional `staticType`.

**Comment by incendial** [comment]
@vHanda hi, do you still need this rule? I've opened a PR with the implementation and I think we'll release it in a week or so. Could you help us with running it on your codebase and providing feedback? If so, we can release it earlier in dev version.

**Comment by incendial** [comment]
Available in 4.2.0 release.

**Comment by vHanda** [comment]
Hi. I just tried it out.

It is wonderful. It has already caught a number of bugs. Thank you!!

**Comment by incendial** [comment]
Happy to hear that! 🎉
